File: packages/fangyuan-rule-java/src/main/java/com/olymtech/ConcurrentRule.java
package com.olymtech;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class ConcurrentRule<T> implements Rule<List<T>> {
    private Map<String, Rule<T>> rules;
    private Map<String, Rule<T>> validRules = new HashMap<>();

    ConcurrentRule(Map<String, Rule<T>> rules){
        this.rules = rules;
    }

    @Override
    public boolean evaluate(final Map<String, Object> ctx) {
        this.validRules = this
                .rules
                .entrySet()
                .stream()
                .filter(rule -> rule.getValue().evaluate(ctx))
                .collect(
                        Collectors.toMap(entry -> entry.getKey(), entry -> entry.getValue())
                );
        return this.validRules.size() > 0;
    }

    @Override
    public List<T> execute(){
        return this.validRules.values().stream().map(rule -> rule.execute()).collect(Collectors.toList());
    }
}

File: packages/fangyuan-rule-sdk/src/ConcurrentRule.ts
import { Rule } from "./Rule";

export class ConcurrentRule<T, R = Record<string, any>> implements Rule<T[]> {
  validRules: Rule<T>[] = [];
  constructor(
    private readonly ctx: R,
    private readonly rules: ((ctx: R) => Rule<T>)[]
  ) {}

  evaluate(): boolean {
    this.validRules = this.rules
      .map((ruleCotr) => ruleCotr(this.ctx))
      .filter((rule) => rule.evaluate());

    return this.validRules.length > 0;
  }

  execute(): T[] {
    if (this.validRules.length === 0) {
      throw new Error("请确保规则执行成功再调用");
    }
    return this.validRules.map((rule) => rule.execute());
  }
}
